### PR TITLE
udevd: wait for workers to finish when exiting

### PR DESCRIPTION
**Pull request by mwilck**
This PR is an alternative, simpler approach to #13956.

On some systems with lots of devices, device probing for certain drivers can
take a very long time. If systemd-udevd detects a timeout and kills the worker
running modprobe using SIGKILL, some devices will not be probed, or end up in
unusable state. The --event-timeout option can be used to modify the maximum
time spent in an uevent handler. But if systemd-udevd exits, it uses a
different timeout, hard-coded to 30s, and exits when this timeout expires,
causing all workers to be KILLed by systemd afterwards. In practice, this may
lead to workers being killed after significantly less time than specified with
the event-timeout. This is particularly significant during initrd processing:
systemd-udevd will be stopped by systemd when initrd-switch-root.target is
about to be isolated, which usually happens quickly after finding and mounting
the root FS.

If systemd-udevd is started by PID 1 (i.e. basically always), systemd will
kill both udevd and the workers after expiry of TimeoutStopSec. This is
actually better than the built-in udevd timeout, because it's more transparent
and configurable for users. This way users can avoid the mentioned boot problem
by simply increasing StopTimeoutSec= in systemd-udevd.service.

### CODE REVIEW

**Comment by fbuihuu** [comment]
FWIW I prefer this approach, it's simpler and it relies on existing PID1 stop-timeout logic so there's no need to add a new option.

File: src/udev/udevd.c [hunk 1]
```diff
@@ -810,11 +798,6 @@ static void manager_exit(Manager *manager) {
         manager_kill_workers(manager);
 
         assert_se(sd_event_now(manager->event, CLOCK_MONOTONIC, &usec) >= 0);
```
Comment (yuwata) [inline src/udev/udevd.c]: This line seems not necessary any more.

Comment (mwilck) [reply src/udev/udevd.c]: PR updated and removed this line.

**Comment by yuwata** [comment]
I noticed that udevd may run without systemd. In that case, the timeout may be important, no?

**Comment by keszybz** [comment]
Yeah, debian & friends still run without systemd, esp. in the initramfs. So this needs to be conditionalized. I think something as simple as `if (sd_booted() == 0)` should be enough.

As to whether this or #13956 is better: I prefer this one too. Simplicity is always good. And I would prefer to avoid any need for explicit configuration from users.

**Comment by keszybz** [comment]
Also: maybe rename this to "let systemd handle killing of workers on exit"? "wait forever" sounds scary ;)

**Comment by mwilck** [comment]
@yuwata,

> I noticed that udevd may run without systemd. In that case, the timeout may be important, no?

I am not aware of all possible scenarios udevd would be run in. We do have the event timeout, which *should* ensure that eventually all running workers terminate (which would then allow udevd to quit, too). Thus I *believe* this would be safe, just the wait time would be longer (event timeout is 90s by default, as opposed to 30s exit timeout).

If we want to be absolutely certain not to break any legacy usage of udevd, we should rather go the #13956 route. That way we could set the udev internal timeout to "infinity" when calling it from systemd, while not changing anything for legacy usage.

**Comment by mwilck** [comment]
@keszybz ,

thanks for looking into this.

> Yeah, debian & friends still run without systemd, esp. in the initramfs. So this needs to be conditionalized. I think something as simple as `if (sd_booted() == 0)` should be enough.

Thanks for the hint.

> As to whether this or #13956 is better: I prefer this one too. Simplicity is always good. And I would prefer to avoid any need for explicit configuration from users.

As I said above, user configuration could be avoided by using different defaults for the timeout whether or not udevd was started by systemd. But yeah, I guess `sd_booted()` is the better solution. I will update this PR.

**Comment by mwilck** [comment]
> Yeah, debian & friends still run without systemd, esp. in the initramfs. So this needs to be conditionalized. I think something as simple as `if (sd_booted() == 0)` should be enough.

Thinking about it again, don't you think it would be sufficient for Debian as well to have the workers killed when the event timeout is reached? How much sense does it make, really, to have these two separate timeouts?

Consider the scenario:
 * worker starts at time `t1` and stars some command that hangs, e.g. modprobe
 * udevd receives exit signal at time `t1 + t2` (`t2 > 0`) and sends SIGTERM to the workers, but the hanging program doesn't terminate on SIGTERM.

With the current code, at `t1 + t2 + 30s` udevd exits despite the still running worker because of the exit timeout, and systemd kills the worker with SIGKILL afterwards as part of udevd's process group. Without systemd, the workers will actually become zombies.

With this patch, udevd itself kills the worker with SIGKILL at `t1 + 90s`. Once all workers are gone, it terminates. This is independent on systemd. Exception: if `TimeoutStopSec  <90s`, systemd will kill udevd and its workers earlier.

IMO this is an improvement over the previous behavior even without systemd. Workers can't end up as zombies any more, and we don't have two competing timeouts (with systemd, three). Hardly anyone is aware of the exit timeout, anyway.
I think this is actually an improvement, 
So even without systemd,

**Comment by keszybz** [comment]
I'm lost: isn't the new patch the same apart from a changed commit message?

**Comment by mwilck** [comment]
I've updated this PR just with changes in the commit message and the NEWS file. 

If you believe the potentially increased wait time for udevd to exit on Debian systems is a show stopper, I'm open for re-adding the timeout with a conditional. 

But I believe this is better than using a  conditional, rationale given in the previous two comments.

It seems I've found a bug though - udevd crashes when the event timeout is hit after manager_exit() is called. This is independent of this PR. I'm currently analyzing it.

**Comment by mwilck** [comment]
> I'm lost: isn't the new patch the same apart from a changed commit message?

Exactly. Rationale given [in this comment](https://github.com/systemd/systemd/pull/13961#issuecomment-552845604).

**Comment by mwilck** [comment]
Anyway, please hold on for a while, I'm working on the problem I just encountered.

**Comment by mwilck** [comment]
OK, found and fixed the bug that I'd encountered. Again, note that this bug existed before this PR already. It would happen if a hanging child's timeout expired before the exit timeout expired.
(worker is started and hangs; 80s later udevd catches an exit signal; 10 more seconds later the child's event timeout exipres; udevd kills child with SIGKILL; udevd receives SIGCHLD and dumps core in on_sigchld() as described in the commit message).

The removal of the exit timeout causes this to happen more easily; now it happens always if udevd is sent SIGTERM while workers are still hanging, and udevd is not killed before it kills its children.

Anyway, with the second patch the issue is fixed for good.

**Comment by keszybz** [comment]
LGTM.

File: src/udev/udevd.c [hunk 2]
```diff
@@ -293,6 +293,8 @@ static void manager_free(Manager *manager) {
         if (!manager)
                 return;
 
+        manager->monitor = sd_device_monitor_unref(manager->monitor);
```
Comment (yuwata) [inline src/udev/udevd.c:296]: It's not necessary to assign value here.

Comment (keszybz) [reply src/udev/udevd.c:296]: I think it's reasonable to do it. We call other functions which take `manager` as argument below, and while they don't look at `manager->monitor`, somebody might change something in one of them later, and avoiding a stale pointer just makes the whole thing more robust.

Comment (yuwata) [reply src/udev/udevd.c:296]: OK.

**Pull request merged**